test/OpenTelemetry.Tests/Metrics/MetricTestsBase.cs

```diff
@@ -130,8 +208,106 @@ public static void CheckTagsForNthMetricPoint(List<Metric> metrics, List<KeyValu
         }
     }
 
+    public IDisposable BuildMeterProvider(
+        out MeterProvider meterProvider,
+        Action<MeterProviderBuilder> configure)
+    {
+        if (configure == null)
+        {
+            throw new ArgumentNullException(nameof(configure));
+        }
+
+#if BUILDING_HOSTING_TESTS
+        var host = BuildHost(
+            useWithMetricsStyle: false,
+            configureMeterProviderBuilder: configure,
+            configureServices: services =>
+            {
+                if (this.configuration != null)
+                {
+                    services.AddSingleton(this.configuration);
+                }
+            });
+
+        meterProvider = host.Services.GetService<MeterProvider>();
+
+        return host;
+#else
+        var builder = Sdk.CreateMeterProviderBuilder();
+
+        if (this.configuration != null)
+        {
+            builder.ConfigureServices(services => services.AddSingleton(this.configuration));
+        }
+
+        configure(builder);
+
+        return meterProvider = builder.Build();
+#endif
+    }
+
     internal static Exemplar[] GetExemplars(MetricPoint mp)
     {
         return mp.GetExemplars().Where(exemplar => exemplar.Timestamp != default).ToArray();
     }
+
+#if BUILDING_HOSTING_TESTS
+    public sealed class HostingMeterProviderBuilder : MeterProviderBuilderBase
+    {
+        public HostingMeterProviderBuilder(IServiceCollection services)
+            : base(services)
+        {
+        }
+
+        public override MeterProviderBuilder AddMeter(params string[] names)
+        {
+            return this.ConfigureServices(services =>
+            {
+                foreach (var name in names)
+                {
+                    // Note: The entire purpose of this class is to use the
+                    // IMetricsBuilder API to enable Metrics and NOT the
+                    // traditional AddMeter API.
+                    services.AddMetrics(builder => builder.EnableMetrics(name));
+                }
+            });
+        }
+
+        public MeterProviderBuilder AddSdkMeter(params string[] names)
+        {
+            return base.AddMeter(names);
+        }
+    }
+
+    private sealed class MetricsSubscriptionManagerCleanupHostedService : IHostedService, IDisposable
+    {
+        private readonly object metricsSubscriptionManager;
+
+        public MetricsSubscriptionManagerCleanupHostedService(IServiceProvider serviceProvider)
+        {
+            this.metricsSubscriptionManager = serviceProvider.GetService(
+                typeof(ConsoleMetrics).Assembly.GetType("Microsoft.Extensions.Diagnostics.Metrics.MetricsSubscriptionManager"));
```

Comment:
nit: We could add the dependency to this package for the test projects.

Reply:
The dependency is already there (via `Microsoft.Extensions.Hosting`). This is hack to fix https://github.com/dotnet/runtime/issues/94434. Hoping it can go away in .NET 9.